mkosi/run.py

```diff
@@ -0,0 +1,313 @@
+import ctypes
+import ctypes.util
+import multiprocessing
+import os
+import pwd
+import shlex
+import signal
+import subprocess
+import sys
+import traceback
+from pathlib import Path
+from types import TracebackType
+from typing import Any, Callable, Iterable, Mapping, Optional, Sequence, Type, TypeVar
+
+from mkosi.backend import MkosiState
+from mkosi.log import ARG_DEBUG, MkosiPrinter, die
+from mkosi.types import _FILE, CompletedProcess, PathString, Popen
+
+CLONE_NEWNS = 0x00020000
+CLONE_NEWUSER = 0x10000000
+
+SUBRANGE = 65536
+
+T = TypeVar("T")
+
+
+def unshare(flags: int) -> None:
+    libc_name = ctypes.util.find_library("c")
+    if libc_name is None:
+        die("Could not find libc")
+    libc = ctypes.CDLL(libc_name, use_errno=True)
+
+    if libc.unshare(ctypes.c_int(flags)) != 0:
+        e = ctypes.get_errno()
+        raise OSError(e, os.strerror(e))
+
+
+def read_subrange(path: Path) -> int:
+    uid = str(os.getuid())
+    try:
+        user = pwd.getpwuid(os.getuid()).pw_name
+    except KeyError:
+        user = None
+
+    for line in path.read_text().splitlines():
+        name, start, count = line.split(":")
+
+        if name == uid or name == user:
+            break
+    else:
+        die(f"No mapping found for {user or uid} in {path}")
+
+    if int(count) != SUBRANGE:
+        die(f"subuid/subgid range length must be {SUBRANGE}, got {count} for {user or uid} from line '{line}'")
+
+    return int(start)
+
+
+def become_root() -> tuple[int, int]:
```

Comment:
```suggestion
def become_root() -> tuple[int, int]:
    """Set up a new user namespace mapping using /etc/subuid and /etc/subgid.
    
    The current user will be mapped to root and 65436 will be mapped to the UID/GID of the invoking user. 
    The other IDs will bemapped through.
    
    The function returns the UID-GID pair of the invoking user in the namespace (65436, 65436).
    """
```
Shame docstrings can't be f-strings, Would be nice if the docs could be self-updating.